galsim/utilities.py

```diff
@@ -296,6 +296,25 @@ def _convertPositions(pos, units, func):
 
     return pos
 
+def _lin_approx_err(x, f, i):
+    """Error as \int abs(f(x) - approx(x)) when using ith data point to make piecewise linear
+    approximation."""
+    xleft, xright = x[:i+1], x[i:]
+    fleft, fright = f[:i+1], f[i:]
+    xi, fi = x[i], f[i]
+    mleft = (fi-f[0])/(xi-x[0])
+    mright = (f[-1]-fi)/(x[-1]-xi)
+    f2left = f[0]+mleft*(xleft-x[0])
+    f2right = fi+mright*(xright-xi)
+    return np.trapz(np.abs(fleft-f2left), xleft), np.trapz(np.abs(fright-f2right), xright)
+
+def _lin_approx_split(x, f):
+    """Optimally split a tabulated function into two-part piecewise linear approximation.
+    """
+    errs = [_lin_approx_err(x, f, i) for i in xrange(1, len(x)-1)]
```

Comment:
I got the O(N) L2 error search to work (with one slight alteration to optimize the integral instead of the sum).  This is definitely a big improvement in speed -- bringing the amount of time needed to thin the LSST bandpasses down from 80s to 2s on my laptop.  The results in accuracy are more complicated.  For a reminder, here's the results of running Jim's script for the slow (but optimal) O(N^2) split point search:

![color_comp](https://cloud.githubusercontent.com/assets/3650485/14802423/86677c34-0b05-11e6-8aba-57f226d979e6.png)

And then here's the result for the O(N) not-quite-optimal split point search:

![default_color_comp](https://cloud.githubusercontent.com/assets/3650485/14802432/a40f50a4-0b05-11e6-8c20-6707a4376fe4.png)

The approximate search doesn't do quite as good of a job in optimizing \int abs(f(x) - approx(x)).  This is despite the fact that the stopping criterion is the same for both algorithms.  I dug a little deeper and looked at the specific combinations of bandpass and SED that were being problematic in the u-band.  In the following figure, the blue shaded region is the u-band throughput, the red shaded region is one of the problematic SEDs, and the green&blue lines are the original and thinned products of these (all normalized to unit maximum).  The red line is the difference between the thinned and unthinned product.

![u-band](https://cloud.githubusercontent.com/assets/3650485/14802500/152e47f4-0b06-11e6-9fc0-413d4a0fb85c.png)

The problematic wavelengths seem to all be due to light leaks.  I can't quite wrap my head around it, but I think this might be something like the L2 search favoring fixing a narrow but more- discrepant region vs a wide but less-discrepant region compared to the L-inf search.

In the end, I think maybe one more keyword with a default, something like `exact_search=False`, may be appropriate, together with a warning in the documentation that the thinning target is only accurate for photometry with respect to a flat (in photons/unit-wavelength) SED.